book/src/state_specific_systems/how_to_define_state_dispatcher_external.md

```diff
@@ -1,156 +1,277 @@
+# }
+# 
+# impl<'a, 'b> SimpleState for CustomDispatcherState<'a, 'b> {}
+# 
+# impl<'a, 'b> CustomDispatcherState<'a, 'b> {
+#     fn new(dispatcher_builder: DispatcherBuilder<'a, 'b>) -> Self {
+#         Self {
+#             dispatcher_builder: Some(dispatcher_builder),
+#             dispatcher: None,
+#         }
+#     }
+# }
+```
 
-        fn update(&mut self, data: StateData<'_, GameData<'a, 'b>>) -> Trans<GameData<'a, 'b>, E> {
-            data.data.update(&data.world);
-            self.dispatcher.as_mut().unwrap().dispatch(&data.world.res);
+This enables us to create the `State` and register `System`s as mentioned above. Now we have to actually build the `Dispatcher` within the `State`. `System` resources are setup when the `Dispatcher` is built. The `System#setup(..)` method needs access to the `World`'s resources, which is made available to `State#on_start(..)`. Therefore, we need to defer building the actual `Dispatcher` until this method is called.
 
-            Trans::Pop
+```rust,edition2018,no_run,noplaypen
+# extern crate amethyst;
+# 
+# use amethyst::{
+#     ecs::prelude::*,
+#     prelude::*,
+# };
+# 
+# /// Builder for the `CustomDispatcherState`.
+# ///
+# /// This allows you to specify which `System`s you want to run within the state.
+# pub struct CustomDispatcherStateBuilder<'a, 'b> {
+#     /// The `State` specific `DispatcherBuilder`.
+#     dispatcher_builder: DispatcherBuilder<'a, 'b>,
+# }
+# 
+# impl<'a, 'b> CustomDispatcherStateBuilder<'a, 'b> {
+#     /// Returns a new `CustomDispatcherStateBuilder`.
+#     pub fn new() -> Self {
+#         CustomDispatcherStateBuilder {
+#             dispatcher_builder: DispatcherBuilder::new(),
+#         }
+#     }
+# 
+#     /// Registers a `System` with the dispatcher builder.
+#     ///
+#     /// # Parameters
+#     ///
+#     /// * `system`: `System` to register.
+#     /// * `name`: Name to register the system with, used for dependency ordering.
+#     /// * `deps`: Names of systems that must run before this system.
+#     pub fn with<Sys>(mut self, system: Sys, name: &str, deps: &[&str]) -> Self
+#         where Sys: for<'c> System<'c> + Send + 'a,
+#     {
+#         self.dispatcher_builder.add(system, name, deps);
+#         self
+#     }
+# 
+#     /// Builds and returns the `CustomDispatcherState`.
+#     pub fn build(self) -> CustomDispatcherState<'a, 'b> {
+#         CustomDispatcherState::new(self.dispatcher_builder)
+#     }
+# }
+# 
+# pub struct CustomDispatcherState<'a, 'b> {
+#     /// The `State` specific `DispatcherBuilder`. This will be used to build the actual dispatcher.
+#     dispatcher_builder: Option<DispatcherBuilder<'a, 'b>>,
+# 
+#     /// The `State` specific `Dispatcher`, containing `System`s only relevant for this `State`.
+#     dispatcher: Option<Dispatcher<'a, 'b>>,
+# }
+# 
+impl<'a, 'b> SimpleState for CustomDispatcherState<'a, 'b> {
+    fn on_start(&mut self, data: StateData<GameData>) {
+        if self.dispatcher.is_none() {
```

Comment:
Isn't that a bit redundant? If we are calling on_start, I'd say it is fairly certain the dispatcher is not defined yet. It would make the code easier to understand in my opinion.